lib/tbot/tbot.go

```diff
@@ -569,3 +575,40 @@ func (p *proxyPingCache) ping(ctx context.Context) (*webclient.PingResponse, err
 
 	return p.cachedValue, nil
 }
+
+type alpnConnUpgradeParams struct {
+	addr     string
+	insecure bool
+}
+
+type alpnProxyConnUpgradeRequiredCache struct {
+	botCfg *config.BotConfig
+	log    *slog.Logger
+
+	mu    sync.RWMutex
+	cache map[alpnConnUpgradeParams]bool
+}
+
+func (a *alpnProxyConnUpgradeRequiredCache) isUpgradeRequired(ctx context.Context, addr string, insecure bool) (bool, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.cache == nil {
+		a.cache = make(map[alpnConnUpgradeParams]bool)
+	}
+	v, ok := a.cache[alpnConnUpgradeParams{addr, insecure}]
+	if ok {
+		return v, nil
+	}
+
+	a.log.DebugContext(ctx, "Testing ALPN upgrade necessary", "addr", addr, "insecure", insecure)
+	v = client.IsALPNConnUpgradeRequired(ctx, addr, insecure)
+	a.log.DebugContext(ctx, "Tested ALPN upgrade necessary", "addr", addr, "insecure", insecure, "result", v)
+	if ctx.Err() != nil {
+		// Check for case where false is returned because client canceled ctx.
+		// We don't want to cache this result.
+		return v, ctx.Err()
+	}
```

Comment:
```suggestion
	if err := ctx.Err(); err != nil {
		// Check for case where false is returned because client canceled ctx.
		// We don't want to cache this result.
		return v, err
	}
```

or maybe even

```suggestion
	if err := ctx.Err(); err != nil {
		// Check for case where false is returned because client canceled ctx.
		// We don't want to cache this result.
		return v, trace.Wrap(err)
	}
```